hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/TestContainerReplicationUnderPlacementPolicy.java

```diff
@@ -0,0 +1,178 @@
+      for (DatanodeDetails node : chosenNodes) {
+        candidateNodes.add(node);
+        numProvidedNode--;
+      }
+    }
+
+    for (int i = 0; i < numProvidedNode; i++) {
+      candidateNodes.add(randomDatanodeDetails());
+    }
+
+    conf = new OzoneConfiguration();
+    containerManager =
+        Mockito.mock(ContainerManager.class);
+    eventQueue = new EventQueue();
+    containerStateManager = new ContainerStateManager(conf);
+
+    Mockito.when(containerManager.getContainerIDs())
+        .thenAnswer(invocation -> containerStateManager.getAllContainerIDs());
+
+    Mockito.when(containerManager.getContainer(Mockito.any(ContainerID.class)))
+        .thenAnswer(invocation -> containerStateManager
+            .getContainer((ContainerID)invocation.getArguments()[0]));
+
+    Mockito.when(containerManager.getContainerReplicas(
+        Mockito.any(ContainerID.class)))
+        .thenAnswer(invocation -> containerStateManager
+            .getContainerReplicas((ContainerID)invocation.getArguments()[0]));
+
+    datanodeCommandHandler = new DatanodeCommandHandler();
+    eventQueue.addHandler(SCMEvents.DATANODE_COMMAND, datanodeCommandHandler);
+
+    SCMNodeMetric scmNodeMetric = Mockito.mock(SCMNodeMetric.class);
+    Mockito.when(scmNodeMetric.get()).thenReturn(new SCMNodeStat(1, 0, 1));
+
+    nodeManager =  Mockito.mock(NodeManager.class);
+    Mockito.when(nodeManager.getNodes(HddsProtos.NodeState.HEALTHY))
+        .thenReturn(candidateNodes);
+
+    Mockito.when(nodeManager.getNodeStat(Mockito.any(DatanodeDetails.class)))
+        .thenReturn(scmNodeMetric);
+
+    containerPlacementPolicy = new SCMContainerPlacementRandom(nodeManager,
+        conf, null, true, null);
+
+    replicationManager = new ReplicationManager(
+        new ReplicationManagerConfiguration(),
+        containerManager,
+        containerPlacementPolicy,
+        eventQueue,
+        new LockManager<>(conf));
+    replicationManager.start();
+    Thread.sleep(100L);
+
+    final ContainerInfo container = getContainer(LifeCycleState.CLOSED);
+    final ContainerID id = container.containerID();
+    containerStateManager.loadContainer(container);
+
+    // One replica in CLOSED state
+    DatanodeDetails node = (chosenNodes != null && !chosenNodes.isEmpty())
+        ? chosenNodes.get(0): randomDatanodeDetails();
+    final Set<ContainerReplica> replicas = getReplicas(id, State.CLOSED, node);
+
+    for (ContainerReplica replica : replicas) {
+      containerStateManager.updateContainerReplica(id, replica);
+    }
+
+    replicationManager.processContainersNow();
+    // Wait for EventQueue to call the event handler
+    Thread.sleep(100L);
+
+    final int currentReplicateCommandCount = datanodeCommandHandler
+        .getInvocationCount(SCMCommandProto.Type.replicateContainerCommand);
+    Assert.assertEquals(expectedCount, currentReplicateCommandCount);
+
+    containerStateManager.close();
+    replicationManager.stop();
+  }
+}
```

Comment:
there is no newline here, but as the final comment, with the suggested extractions and reorg, the test class would be someting like this:
{
variables;
testcase1;
testcase2;
testcase3;
private methods;
}

The test cases would look like something like this:
{
replicationmgr = setupReplicationMgr(choosenNodes, numProvidedNode) //numProvidedNode may not needed
updateReplicas()

verification code;
}

where setupReplicationMgr may get the DataNodeCommandHandler from outside as well, and can do what is done to setup the mgr like:
{
collect candidate nodes;
setup the nodeManager and the policy based on candidates
initialize the eventqueue
instantiate and start the mgr
}

With that I believe it would be clearer what is happening where, and if later someone wants to add more tests, or more assertions, it would be easier to extend the logic also. Note that these are just my beliefs, for which I have an argument only if someone says that my pseudo code would be better to understand in real code comparing to the current code:) And I know it is just a test, and maybe I am too rigorous, but small steps also matter.